src/engine/SpatialJoin.cpp

```diff
@@ -0,0 +1,400 @@
+    if (addDistToResult) {
+      return widthChildren + 1;
+    } else {
+      return widthChildren;
+    }
+  } else {
+    return 1;  // dummy return, as the class does not have its children yet
+  }
+}
+
+// ____________________________________________________________________________
+size_t SpatialJoin::getCostEstimate() {
+  if (childLeft_ && childRight_) {
+    return childLeft_->getSizeEstimate() * childRight_->getSizeEstimate();
+  }
+  return 1;  // dummy return, as the class does not have its children yet
+}
+
+// ____________________________________________________________________________
+uint64_t SpatialJoin::getSizeEstimateBeforeLimit() {
+  if (childLeft_ && childRight_) {
+    return childLeft_->getSizeEstimate() * childRight_->getSizeEstimate();
+  }
+  return 1;  // dummy return if not both children are added
+}
+
+// ____________________________________________________________________________
+float SpatialJoin::getMultiplicity(size_t col) {
+  auto getDistinctness = [](std::shared_ptr<QueryExecutionTree> child,
+                            ColumnIndex ind) {
+    auto size = (u_int)child->getSizeEstimate();
+    auto multiplicity = child->getMultiplicity(ind);
+    return size / multiplicity;
+  };
+
+  if (childLeft_ && childRight_) {
+    auto varColsLeftMap = childLeft_->getVariableColumns();
+    auto varColsRightMap = childRight_->getVariableColumns();
+    auto varColMap = computeVariableToColumnMap();
+    auto colVarMap = copySortedByColumnIndex(varColMap);
+    Variable var = colVarMap.at(col).first;
+    auto left = varColsLeftMap.find(var);
+    auto right = varColsRightMap.find(var);
+
+    if (var.name() == nameDistanceInternal) {
+      // as each distance is very likely to be unique (even if only after
+      // a few decimal places), no multiplicities are assumed
+      return 1;
+    } else if (left != varColsLeftMap.end()) {
+      auto distinctnessChild =
+          getDistinctness(childLeft_, left->second.columnIndex_);
+      return getSizeEstimate() / distinctnessChild;
+    } else if (right != varColsRightMap.end()) {
+      auto distinctnessChild =
+          getDistinctness(childRight_, right->second.columnIndex_);
+      return getSizeEstimate() / distinctnessChild;
+    } else {
+      AD_FAIL();  // this should not be reachable
+    }
+    return 1;  // to prevent compiler warning
+  } else {
+    return 1;  // dummy return, as the class does not have its children yet
+  }
+}
+
+// ____________________________________________________________________________
+bool SpatialJoin::knownEmptyResult() {
+  if (childLeft_) {
+    if (childLeft_->knownEmptyResult()) {
+      return true;
+    }
+  }
+
+  if (childRight_) {
+    if (childRight_->knownEmptyResult()) {
+      return true;
+    }
+  }
+
+  // return false if either both children don't have an empty result, only one
+  // child is available, which doesn't have a known empty result or neither
+  // child is available
+  return false;
```

Comment:
your implementation is not the same. Consider the following situation: the SpatialJoin has only one child (therefore isConstructed returns false and your return value is false). The one child however has a knownEmptyResult of true. Then the function should return true, because no matter the size of the other child, the SpatialJoin will result in an empty result

Reply:
`return ((childLeft_ && childLeft->knownE()) || (childRight && childR->knownE()));